src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs

```diff
@@ -303,7 +305,7 @@ public CompilationAnalysisContext(Compilation compilation, AnalyzerOptions optio
         /// <param name="diagnostic"><see cref="Diagnostic"/> to be reported.</param>
         public void ReportDiagnostic(Diagnostic diagnostic)
         {
-            DiagnosticAnalysisContextHelpers.VerifyArguments(diagnostic);
+            DiagnosticAnalysisContextHelpers.VerifyArguments(diagnostic, _isSupportedDiagnostic);
             lock (_reportDiagnostic)
```

Comment:
The constructor is only invoked from the within the IDE and compiler analyzer drivers, and #1285 tracks internalizing these constructors.

We need to lock the reportDiagnostic callback as the analyzer might decide to kick off multiple threads and report diagnostics from different threads simultaneously. We can probably guarantee that all analyzer hosts use thread safe diagnostic bags and the Add delegate that they provide for reported diagnostics is thread safe. However, this code is not maintainable and likely to introduce future race conditions. The only foolproof approach here is to make ReportDiagnostic itself intrinsically thread-safe by using a lock.